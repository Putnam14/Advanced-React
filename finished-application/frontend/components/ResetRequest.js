import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';

const REQUEST_RESET_MUTATION = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      id
    }
  }
`;

class ResetRequest extends React.Component {
  state = {
    email: '',
  };

  render() {
    return (
      <Mutation mutation={REQUEST_RESET_MUTATION} variables={this.state}>
        {(resetMutation, { loading, error, called }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              await resetMutation();
            }}
            data-test="ResetRequest"
          >
            <Error error={error} />
            {!error && called && !loading && <p>Success! Check Your Email!</p>}
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="email">
                Email
                <input
                  value={this.state.email}
                  onChange={e => this.setState({ email: e.target.value })}
                  name="email"
                  type="text"
                  placeholder="email"
                />
              </label>

              <button type="submit">Request Reset!</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default ResetRequest;
export { REQUEST_RESET_MUTATION };
