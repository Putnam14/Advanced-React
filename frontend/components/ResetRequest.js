import React from 'react';
import { Mutation } from 'react-apollo';
import { REQUEST_RESET_MUTATION } from '../queries/queries';
import Form from './styles/Form';
import Error from './ErrorMessage';

class ResetRequest extends React.Component {
  state = {
    email: '',
  };

  render() {
    return (
      <Mutation mutation={REQUEST_RESET_MUTATION} variables={this.state}>
        {(resetMutation, { loading, error, called, data }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              // I wanna spy on this
              const res = await resetMutation();
              console.log(res);
            }}
            data-test="ResetRequest"
          >
            <Error error={error} />
            {!error && called && !loading && <p>Success! Check Your Email!</p>}
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="email">
                Email
                <input
                  value={this.state.email}
                  onChange={e => this.setState({ email: e.target.value })}
                  name="email"
                  type="text"
                  placeholder="email"
                />
              </label>

              <button type="submit">Request Reset!</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default ResetRequest;
