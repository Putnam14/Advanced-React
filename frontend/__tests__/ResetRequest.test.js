import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ApolloProvider } from 'react-apollo';
import ResetRequest from '../components/ResetRequest';
import mountOptions from './mockMang';

describe('<ResetRequest/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(<ResetRequest />, mountOptions);
    const Form = wrapper.find("Form[data-test='ResetRequest']");
    expect(toJSON(Form)).toMatchSnapshot();
  });

  it('calls the mutation', async () => {
    const wrapper = mount(<ResetRequest />, mountOptions);
    const mutation = wrapper.find('Mutation').instance();
    const input = wrapper.find('input[name="email"]');
    input.simulate('change', { target: { name: 'email', value: 'wesbos@gmail.com' } });
    // Create a spy function on mutate
    mutation.client.mutate = jest.fn();
    mutation.forceUpdate();
    wrapper.find('Form').simulate('submit');
    expect(mutation.client.mutate.mock.calls[0][0].variables).toEqual({
      email: 'wesbos@gmail.com',
    });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders errors', () => {
    const wrapper = mount(<ResetRequest />, mountOptions);
    const mutation = wrapper.find('Mutation').instance();
    mutation.setState({ error: { message: 'Shit!' } });
    wrapper.update();
    const error = wrapper.find('DisplayError');
    expect(error).toHaveLength(1);
    expect(toJSON(error)).toMatchSnapshot();
  });
});
