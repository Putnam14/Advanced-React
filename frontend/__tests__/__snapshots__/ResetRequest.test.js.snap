// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetRequest/> calls the mutation 1`] = `
<ResetRequest>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "requestReset",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "email",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "requestReset",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "email",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 97,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "email": "wesbos@gmail.com",
      }
    }
  >
    <Form
      data-test="ResetRequest"
      onSubmit={[Function]}
    >
      <form
        className="sc-bdVaJa LHejR"
        data-test="ResetRequest"
        onSubmit={[Function]}
      >
        <DisplayError />
        <fieldset
          aria-busy={true}
          disabled={true}
        >
          <label
            htmlFor="email"
          >
            Email
            <input
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value="wesbos@gmail.com"
            />
          </label>
          <button
            type="submit"
          >
            Request Reset!
          </button>
        </fieldset>
      </form>
    </Form>
  </Mutation>
</ResetRequest>
`;

exports[`<ResetRequest/> renders and matches snapshot 1`] = `
<Form
  data-test="ResetRequest"
  onSubmit={[Function]}
>
  <form
    className="sc-bdVaJa LHejR"
    data-test="ResetRequest"
    onSubmit={[Function]}
  >
    <DisplayError />
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <label
        htmlFor="email"
      >
        Email
        <input
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Reset!
      </button>
    </fieldset>
  </form>
</Form>
`;

exports[`<ResetRequest/> renders errors 1`] = `
<DisplayError
  error={
    Object {
      "message": "Shit!",
    }
  }
>
  <styled.div>
    <div
      className="sc-bwzfXH flKeYn"
    >
      <p>
        <strong>
          Shoot!
        </strong>
        Shit!
        <button>
          Try Again
        </button>
      </p>
    </div>
  </styled.div>
</DisplayError>
`;
